### PR TITLE
fixed: offset replay mechanism

### PR DESCRIPTION
**Pull request by devansh42**
1. previously we were seeking for byte offset but now we'll seek entry offset
2. changed wal file naming convention to ease offset replaying

### CODE REVIEW

File: wal.go [hunk 1]
```diff
@@ -157,34 +164,78 @@ func (wal *WriteAheadLog) deleteOldestSegment() error {
...
+}
+
 func (wal *WriteAheadLog) Replay(offset int64, f func([]byte) error) error {
 	logFiles, err := filepath.Glob(wal.logFileName + "*")
 	if err != nil {
 		return err
 	}
-
-	for _, logFile := range logFiles {
+	index, startingOffset, err := wal.findStartingLogFile(offset, logFiles)
```
Comment (aarthikrao) [inline wal.go:234]: Can we use sort the files and then start reading. It will be simpler.

File: wal.go [hunk 2]
```diff
@@ -157,34 +164,78 @@ func (wal *WriteAheadLog) deleteOldestSegment() error {
...
+	for i, logFile := range logFiles[index:] {
 		file, err := os.Open(logFile)
 		if err != nil {
 			return err
 		}
 		defer file.Close()
 
-		if _, err := file.Seek(offset, io.SeekStart); err != nil {
+		if i > 0 {
+			if _, err := file.Seek(0, io.SeekStart); err != nil {
```
Comment (aarthikrao) [inline wal.go]: I'm a little unclear on how this logic works. Can you please add comments.

**Comment by aarthikrao** [comment]
Instead of fixing the byte size of the file, we can set a limit that we will store maximum 1000 lines in one segment. We can name the file with the starting segment. This will help us read from a particular monotonically incrementing counter for the entire wal.( the counter will increment once for every entry into the wal)

**Comment by devansh42** [comment]
The problem, here is the random seeks we need to make to go to exact
offset. With 1000 lines or entries max solution we would have the same
issue that is jumping on a given offset

**Comment by devansh42** [comment]
* Added buffer reader, it would fasten the process of offset seeking, as while seeking record by record we are going to read the same memory page multiple times. Reading one page at a time while reduce redundant page reads as well.

**Review by aarthikrao — APPROVED** [review]

**Pull request merged**